token/lib/_contractHelper.js

```diff
@@ -0,0 +1,212 @@
+   * @params {string} networkId - Network ID.
+   * @returns {object} - Promise that resolves to contract object.
+   */
+  web3(networkId) {
+    const provider = this.config.providers[networkId]
+    return new Web3(provider)
+  }
+
+  /**
+   * Changes the owner of the given contract to the given address.
+   * @param {string} networkId - Ethereum network ID.
+   * @param {string} newOwner - Address of the new owner.
+   */
+  async _setOwner(networkId, contract, newOwner) {
+    const sender = await this.defaultAccount(networkId)
+    const newOwnerLower = newOwner.toLowerCase()
+
+    // Pre-contract call validations.
+    if (
+      !this.config.overrideOwnerWhitelist &&
+      !isValidOwner(networkId, newOwner, this.validOwners)
+    ) {
+      throw new Error(`${newOwner} is not a valid owner for the token contract`)
+    }
+    await this.ensureContractOwner(contract, sender)
+    const oldOwner = await contract.methods.owner().call()
+    if (oldOwner.toLowerCase() === newOwnerLower) {
+      throw new Error('old and new owner are the same')
+    }
+
+    const transaction = contract.methods.transferOwnership(newOwner)
+    await this.sendTransaction(networkId, transaction, { from: sender })
+    await withRetries(this.retries, async () => {
+      const ownerAfterTransaction =
+        (await contract.methods.owner().call()).toLowerCase()
+      if (
+        !this.config.multisig &&
+        ownerAfterTransaction !== newOwner.toLowerCase()
+      ) {
+        throw new Error(`New owner should be ${newOwner} but is ${ownerAfterTransaction}`)
+      }
+    }, this.config.verbose)
+  }
+
+  /**
+   * Sends an Ethereum transaction.
+   * @param {string} networkId - Ethereum network ID.
+   * @param {transaction} transaction - These are returned by contract.methods.MyMethod()
+   * @param {Object} opts - Options to be sent along with the transaction.
+   * @returns {Object} - Transaction receipt.
+   */
+  async sendTransaction(networkId, transaction, opts = {}) {
+    const web3 = this.web3(networkId)
+
+    if (!opts.from) {
+      opts.from = await this.defaultAccount(networkId)
+    }
+
+    if (this.config.multisig) {
+      // For multi-sig transactions, we submit the transaction to the multi-sig
+      // wallet instead of the token contract.
+      const contract = await this.contract(networkId)
+      transaction = await this.multiSigTransaction({
+        networkId,
+        sender: opts.from,
+        multiSigWalletAddress: this.config.multisig,
+        contractAddress: contract._address,
+        transaction
+      })
+    }
+
+    if (!opts.gas) {
+      opts.gas = await transaction.estimateGas({ from: opts.from })
+      this.vlog('estimated gas:', opts.gas)
+    }
+
+    // Send the transaction and grab the transaction hash when it's available.
+    this.vlog('sending transaction')
+    let transactionHash
+    transaction.send(opts)
+      .on('transactionHash', (hash) => {
```

Comment:
Enhancement idea: I wonder if it would be simpler to use the ["confirmation" callback](https://web3js.readthedocs.io/en/1.0/web3-eth.html#eth-sendtransaction-return). That callback includes the receipt. As a bonus it would allow us to be extra safe by waiting for N block confirmations to make sure the block indeed made it to the blockchain.

Reply:
For some contract calls, the `receipt` callback never fires for me using rinkeby.infura.io. I'm curious if the `confirmation` callback will. I'll try it sometime.